README.md
![Banner](https://s-christy.com/status-banner-service/svg-rs/banner-slim.svg)

## Overview

This is a Rust library for creating SVG files. The library supports many of the
basic shapes, text, and attributes like style and transform. This library is not
feature complete.

## Examples

<img alt="SVG example" style="" src="./sample/playground.png">

The above was created with this code:

```rust
let mut svg = Svg::new(400.0, 300.0);

let mut path = Path::new(50.0, 0.0);

path.line_to(0.0, 100.0);

path.line_to(100.0, 100.0);

svg.add_element(format!("<path d=\"{}\" />", path).as_str());

let circle = Circle::new(150.0, 50.0, 40.0, "", "fill:black");

svg.add_shape(circle);

svg.add_shape(Rectangle::new(230.0, 30.0, 50.0, 50.0, "", "fill:black"));

let mut rect = Rectangle::new(330.0, 30.0, 50.0, 50.0, "", "fill:black");

rect.rounded(8.0, 8.0);

svg.add_shape(rect);

let line = Line::new(10.0, 110.0, 80.0, 180.0, "", "stroke:black");

svg.add_shape(line);

svg.add_shape(Rectangle::new(130.0, 130.0, 50.0, 50.0, "", "fill:black"));

svg.add_shape(Rectangle::new(
    30.0,
    30.0,
    10.0,
    10.0,
    "translate(200 100) rotate(40 20 40)",
    "fill:red",
));

svg.add_shape(Text::new(
    310.0,
    110.0,
    10.0,
    "",
    "fill:black",
    "Hello, World!",
));

let bezier = Bezier::new(
    40.0,
    240.0,
    80.0,
    240.0,
    40.0,
    280.0,
    80.0,
    280.0,
    "",
    "stroke:black; fill:none; stroke-width:1",
);

svg.add_shape(bezier);
```

## Features

- API for creating common shapes/primitives in SVG
- Support for transforms
- Support for paths and Beziér curves
- Function to write SVG data to file
- Optional style and transform for each element
- Tree-based storage for nested structures
- Test suite

## Usage

```svg
// Create a new Svg object
let mut svg = Svg::new(100.0, 100.0);

// Create shapes
let circle = Circle::new(50.0, 50.0, 10.0, "", "fill:black");

// Add shapes to the object
svg.add_shape(circle);

// Write the data to a file
svg.to_file("some/file/path.svg").unwrap();
```

See the `playground` test in `src/lib.rs` for further example usage.

## License

This work is licensed under the GNU General Public License version 3 (GPLv3).

[<img src="https://s-christy.com/status-banner-service/GPLv3_Logo.svg" width="150" />](https://www.gnu.org/licenses/gpl-3.0.en.html)


CHANGELOG.md
# Changelog

## [1.0.0] - 2023-01-11

### Added

- API for creating common shapes/primitives in SVG
- Support for transforms
- Support for paths and Beziér curves
- Function to write SVG data to file
- Optional style and transform for each element
- Tree-based storage for nested structures
- Test suite

### Changed

- None (initial release)


src/svg.rs
use crate::shape;
use std::fmt;
use std::fs::File;
use std::io::Write;
use trees;
use trees::tr;

fn tree_to_string(node: &trees::Node<Node>) -> String {
    if node.has_no_child() {
        node.data().begin.clone() + node.data().end.clone().as_str()
    } else {
        format!(
            "{}\n{}{}",
            node.data().begin,
            node.iter()
                .fold(String::new(), |s, c| s + &tree_to_string(c) + "\n"),
            node.data().end,
        )
    }
}

#[derive(Clone)]
pub struct Node {
    pub begin: String,
    pub end: String,
}

impl fmt::Display for Node {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{}{}", self.begin, self.end)
    }
}

pub struct Svg {
    width: f32,
    height: f32,
    pub nodes: trees::Tree<Node>,
}

impl Svg {
    #[must_use] pub fn new(width: f32, height: f32) -> Self {
        Self {
            width,
            height,
            nodes: tr(Node {
                begin: String::new()
                    + format!(
                        "<svg viewBox=\"0 0 {} {}\" xmlns=\"http://www.w3.org/2000/svg\">",
                        width, height
                    )
                    .as_str(),
                end: String::new() + "</svg>",
            }),
        }
    }

    pub fn add_element(&mut self, s: &str) {
        self.nodes = self.nodes.clone()
            / tr(Node {
                begin: s.to_string(),
                end: String::new(),
            });
    }

    pub fn add_shape<S: shape::Shape>(&mut self, s: S) {
        self.nodes = self.nodes.clone()
            / tr(Node {
                begin: s.to_svg(),
                end: String::new(),
            });
    }

    pub fn set_width(&mut self, width: f32) {
        self.width = width;
    }

    pub fn set_height(&mut self, height: f32) {
        self.height = height;
    }

    pub fn to_file(&self, path: &str) -> std::io::Result<()> {
        let mut output = File::create(path)?;
        write!(output, "{}", self)
    }
}

impl fmt::Display for Svg {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{}", tree_to_string(&self.nodes))
    }
}


src/path.rs
use std::fmt;

pub struct Path {
    points: Vec<(f32, f32)>,
}

impl Path {
    #[must_use]
    pub fn new(x: f32, y: f32) -> Self {
        Self {
            points: vec![(x, y)],
        }
    }

    pub fn line_to(&mut self, x: f32, y: f32) {
        self.points.push((x, y));
    }
}

impl fmt::Display for Path {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let mut ret = String::new();

        let mut first = true;

        for point in &self.points {
            if first {
                ret += "M ";
                first = false;
            } else {
                ret += "L ";
            }
            ret += point.0.to_string().as_str();
            ret += " ";
            ret += point.1.to_string().as_str();
            ret += " ";
        }

        ret += "Z";
        write!(f, "{}", ret)
    }
}


src/shape.rs
pub trait Shape {
    fn to_svg(&self) -> String;
}

pub struct Circle<'a> {
    x: f32,
    y: f32,
    r: f32,
    style: &'a str,
    transform: &'a str,
}

impl Shape for Circle<'_> {
    fn to_svg(&self) -> String {
        format!(
            "<circle cx=\"{}\" cy=\"{}\" r=\"{}\" \
            transform=\"{}\" style=\"{}\" />",
            self.x, self.y, self.r, self.transform, self.style
        )
    }
}

impl<'a> Circle<'a> {
    #[must_use]
    pub fn new(x: f32, y: f32, r: f32, transform: &'a str, style: &'a str) -> Self {
        Self {
            x,
            y,
            r,
            style,
            transform,
        }
    }
}

pub struct Rectangle<'a> {
    x: f32,
    y: f32,
    rx: f32,
    ry: f32,
    width: f32,
    height: f32,
    style: &'a str,
    transform: &'a str,
}

impl Shape for Rectangle<'_> {
    fn to_svg(&self) -> String {
        format!(
            "<rect x=\"{}\" y=\"{}\" width=\"{}\" height=\"{}\" \
            rx=\"{}\" ry=\"{}\" \
            transform=\"{}\" style=\"{}\" />",
            self.x, self.y, self.width, self.height, self.rx, self.ry, self.transform, self.style
        )
    }
}

impl<'a> Rectangle<'a> {
    #[must_use]
    pub fn new(
        x: f32,
        y: f32,
        width: f32,
        height: f32,
        transform: &'a str,
        style: &'a str,
    ) -> Self {
        Self {
            x,
            y,
            rx: 0.0,
            ry: 0.0,
            width,
            height,
            style,
            transform,
        }
    }

    pub fn rounded(&mut self, rx: f32, ry: f32) {
        self.rx = rx;
        self.ry = ry;
    }
}

pub struct Line<'a> {
    x1: f32,
    y1: f32,
    x2: f32,
    y2: f32,
    style: &'a str,
    transform: &'a str,
}

impl Shape for Line<'_> {
    fn to_svg(&self) -> String {
        format!(
            "<line x1=\"{}\" y1=\"{}\" x2=\"{}\" y2=\"{}\" \
            transform=\"{}\" style=\"{}\" />",
            self.x1, self.y1, self.x2, self.y2, self.transform, self.style
        )
    }
}

impl<'a> Line<'a> {
    #[must_use]
    pub fn new(x1: f32, y1: f32, x2: f32, y2: f32, transform: &'a str, style: &'a str) -> Self {
        Self {
            x1,
            y1,
            x2,
            y2,
            style,
            transform,
        }
    }
}

pub struct Bezier<'a> {
    x1: f32,
    y1: f32,
    x2: f32,
    y2: f32,
    x3: f32,
    y3: f32,
    x4: f32,
    y4: f32,
    style: &'a str,
    transform: &'a str,
}

impl Shape for Bezier<'_> {
    fn to_svg(&self) -> String {
        format!(
            "<path d=\"M {} {} C {} {} \
            {} {} {} {}\" \
            transform=\"{}\" style=\"{}\" />",
            self.x1,
            self.y1,
            self.x2,
            self.y2,
            self.x3,
            self.y3,
            self.x4,
            self.y4,
            self.transform,
            self.style
        )
    }
}

impl<'a> Bezier<'a> {
    #[must_use]
    pub fn new(
        x1: f32,
        y1: f32,
        x2: f32,
        y2: f32,
        x3: f32,
        y3: f32,
        x4: f32,
        y4: f32,
        transform: &'a str,
        style: &'a str,
    ) -> Self {
        Self {
            x1,
            y1,
            x2,
            y2,
            x3,
            y3,
            x4,
            y4,
            style,
            transform,
        }
    }
}

pub struct Polygon<'a> {
    points: &'a str,
    style: &'a str,
    transform: &'a str,
}

impl Shape for Polygon<'_> {
    fn to_svg(&self) -> String {
        format!(
            "<polygon points=\"{}\" \
            transform=\"{}\" style=\"{}\" />",
            self.points, self.transform, self.style
        )
    }
}

impl<'a> Polygon<'a> {
    #[must_use]
    pub fn new(points: &'a str, transform: &'a str, style: &'a str) -> Self {
        Self {
            points,
            style,
            transform,
        }
    }
}


src/lib.rs
pub mod path;
pub mod shape;
pub mod svg;
pub mod text;

#[cfg(test)]
mod tests {
    use super::*;
    use path::Path;
    use shape::Bezier;
    use shape::Circle;
    use shape::Line;
    use shape::Rectangle;
    use svg::Svg;
    use text::Text;

    #[test]
    fn hello_world() {
        let svg = Svg::new(50.0, 50.0);

        svg.to_file("target/hello_world.svg").unwrap();

        assert_eq!(
            svg.to_string(),
            "<svg viewBox=\"0 0 50 50\" \
            xmlns=\"http://www.w3.org/2000/svg\"></svg>"
        );
    }

    #[test]
    fn circle() {
        let mut svg = Svg::new(50.0, 50.0);

        let circle = Circle::new(50.0, 50.0, 40.0, "", "fill:black");

        svg.add_shape(circle);

        svg.to_file("target/circle.svg").unwrap();

        assert_eq!(
            svg.to_string(),
            "<svg viewBox=\"0 0 50 50\" xmlns=\"http://www.w3.org/2000/svg\">\n\
            <circle cx=\"50\" cy=\"50\" r=\"40\" transform=\"\" style=\"fill:black\" />\n\
            </svg>"
        );
    }

    #[test]
    fn rectangle() {
        let mut svg = Svg::new(50.0, 50.0);

        svg.add_shape(Rectangle::new(30.0, 30.0, 10.0, 10.0, "", "fill:black"));

        svg.to_file("target/rectangle.svg").unwrap();

        assert_eq!(
            svg.to_string(),
            "<svg viewBox=\"0 0 50 50\" xmlns=\"http://www.w3.org/2000/svg\">\n\
            <rect x=\"30\" y=\"30\" width=\"10\" height=\"10\" \
            rx=\"0\" ry=\"0\" transform=\"\" style=\"fill:black\" />\n\
            </svg>"
        );
    }

    #[test]
    fn rectangle_rounded() {
        let mut svg = Svg::new(50.0, 50.0);

        let mut rect = Rectangle::new(30.0, 30.0, 10.0, 10.0, "", "fill:black");

        rect.rounded(2.0, 2.0);

        svg.add_shape(rect);

        svg.to_file("target/rectangle_rounded.svg").unwrap();

        assert_eq!(
            svg.to_string(),
            "<svg viewBox=\"0 0 50 50\" xmlns=\"http://www.w3.org/2000/svg\">\n\
            <rect x=\"30\" y=\"30\" width=\"10\" height=\"10\" \
            rx=\"2\" ry=\"2\" transform=\"\" style=\"fill:black\" />\n\
            </svg>"
        );
    }

    #[test]
    fn line() {
        let mut svg = Svg::new(50.0, 50.0);

        let line = Line::new(10.0, 10.0, 20.0, 20.0, "", "stroke:black");

        svg.add_shape(line);

        svg.to_file("target/line.svg").unwrap();

        assert_eq!(
            svg.to_string(),
            "<svg viewBox=\"0 0 50 50\" xmlns=\"http://www.w3.org/2000/svg\">\n\
            <line x1=\"10\" y1=\"10\" x2=\"20\" y2=\"20\" \
            transform=\"\" style=\"stroke:black\" />\n\
            </svg>"
        );
    }

    #[test]
    fn bezier() {
        let mut svg = Svg::new(50.0, 50.0);

        let bezier = Bezier::new(
            10.0,
            10.0,
            20.0,
            10.0,
            10.0,
            20.0,
            20.0,
            20.0,
            "",
            "stroke:black; fill:none; stroke-width:.1",
        );

        svg.add_shape(bezier);

        svg.to_file("target/bezier.svg").unwrap();

        assert_eq!(
            svg.to_string(),
            "<svg viewBox=\"0 0 50 50\" xmlns=\"http://www.w3.org/2000/svg\">\n\
            <path d=\"M 10 10 C 20 10 10 20 20 20\" \
            transform=\"\" style=\"stroke:black; fill:none; \
            stroke-width:.1\" />\n</svg>",
        );
    }

    #[test]
    fn text() {
        let mut svg = Svg::new(50.0, 50.0);

        svg.add_shape(Text::new(
            10.0,
            10.0,
            2.0,
            "",
            "fill:black",
            "Hello, World!",
        ));

        assert_eq!(
            svg.to_string(),
            "<svg viewBox=\"0 0 50 50\" xmlns=\"http://www.w3.org/2000/svg\">\n\
            <text x=\"10\" y=\"10\" font-size=\"2\" \
            transform=\"\" style=\"fill:black\">Hello, World!</text>\n\
            </svg>",
        );
    }

    #[test]
    fn double_circle() {
        let mut svg = Svg::new(100.0, 100.0);

        svg.add_shape(Circle::new(50.0, 50.0, 40.0, "", "fill:black"));
        svg.add_shape(Circle::new(50.0, 50.0, 40.0, "", "fill:black"));

        svg.to_file("target/double_circle.svg").unwrap();

        assert_eq!(
            svg.to_string(),
            "<svg viewBox=\"0 0 100 100\" xmlns=\"http://www.w3.org/2000/svg\">\n\
            <circle cx=\"50\" cy=\"50\" r=\"40\" transform=\"\" style=\"fill:black\" />\n\
            <circle cx=\"50\" cy=\"50\" r=\"40\" transform=\"\" style=\"fill:black\" />\n\
            </svg>"
        );
    }

    #[test]
    fn transform() {
        let mut svg = Svg::new(100.0, 100.0);

        svg.add_shape(Rectangle::new(30.0, 30.0, 10.0, 10.0, "", "fill:black"));

        svg.add_shape(Rectangle::new(
            30.0,
            30.0,
            10.0,
            10.0,
            "rotate(40 20 40)",
            "fill:red",
        ));

        assert_eq!(
            svg.to_string(),
            "<svg viewBox=\"0 0 100 100\" xmlns=\"http://www.w3.org/2000/svg\">\n\
            <rect x=\"30\" y=\"30\" width=\"10\" height=\"10\" rx=\"0\" ry=\"0\" \
            transform=\"\" style=\"fill:black\" />\n\
            <rect x=\"30\" y=\"30\" width=\"10\" height=\"10\" rx=\"0\" ry=\"0\" \
            transform=\"rotate(40 20 40)\" style=\"fill:red\" />\n\
            </svg>"
        );
    }

    #[test]
    fn playground() {
        let mut svg = Svg::new(400.0, 300.0);

        let mut path = Path::new(50.0, 0.0);

        path.line_to(0.0, 100.0);

        path.line_to(100.0, 100.0);

        svg.add_element(format!("<path d=\"{}\" />", path).as_str());

        let circle = Circle::new(150.0, 50.0, 40.0, "", "fill:black");

        svg.add_shape(circle);

        svg.add_shape(Rectangle::new(230.0, 30.0, 50.0, 50.0, "", "fill:black"));

        let mut rect = Rectangle::new(330.0, 30.0, 50.0, 50.0, "", "fill:black");

        rect.rounded(8.0, 8.0);

        svg.add_shape(rect);

        let line = Line::new(10.0, 110.0, 80.0, 180.0, "", "stroke:black");

        svg.add_shape(line);

        svg.add_shape(Rectangle::new(130.0, 130.0, 50.0, 50.0, "", "fill:black"));

        svg.add_shape(Rectangle::new(
            30.0,
            30.0,
            10.0,
            10.0,
            "translate(200 100) rotate(40 20 40)",
            "fill:red",
        ));

        svg.add_shape(Text::new(
            310.0,
            110.0,
            10.0,
            "",
            "fill:black",
            "Hello, World!",
        ));

        let bezier = Bezier::new(
            40.0,
            240.0,
            80.0,
            240.0,
            40.0,
            280.0,
            80.0,
            280.0,
            "",
            "stroke:black; fill:none; stroke-width:1",
        );

        svg.add_shape(bezier);

        //println!("\n\n{}\n\n", svg);

        //svg.to_file("out.svg").unwrap();

        svg.to_file("target/playground.svg").unwrap();

        assert_eq!(1, 1);
    }
}
